File: sscal.c
/*
 * sscal.c
 *
 * This routine is the C version of the FORTRAN routine used in LINPACK
 * but converted to C by f2c. The purpose of this conversion is to allow
 * significant changes to be made to the code while maintaining it's
 * API to the other FORTRAN routines in which it might be linked.
 *
 * The goal of this is to make these routines capable of using SMP machines
 * so that a more powerful machine can solve the problem faster. Whether
 * or not this is reached, we do not know, but that's the goal.
 *
 * The first pass at this routine is to make use of threads and split up
 * the problem into several threads so that the operating system can
 * put the different threads on different processors.
 */

/* 	System Headers 			*/

/* 	Third-Party Headers 	*/

/* 	Other Headers 			*/

/* 	Forward Declarations 	*/

/* 	Public Data Types 		*/
/*
 * Because we need to be able to build this for 32-bit and 64-bit
 * versions, I want to be able to typedef the integer here so that
 * the value coming in from the FORTRAN code matches what we will
 * use here. Without this, we'd have a mess on the conversions.
 */
#if defined(__x86_64__) || defined(__ppc64__)
typedef int f_int;
#else
typedef long int f_int;
#endif

/* 	Public Data Constants 	*/


/* ----------------------------------------------------------------------------
 *
 *   computer            - cdc/single
 *
 *   latest revision     - january 1, 1978
 *
 *   purpose             - compute a single precision constant
 *                         times a single precision vector
 *
 *   usage               - call sscal (n,sa,sx,incx)
 *
 *   arguments    n      - length of vector x. (input)
 *                sa     - real scalar. (input)
 *                sx     - real vector of length max(n*iabs(incx),1).
 *                         (input)
 *                         sscal replaces x(i) with sa*x(i) for
 *                         i=1,...,n.
 *                         x(i) refers to a specific element of sx.
 *                         see incx argument description.
 *                incx   - displacement between elements of sx. (input)
 *                         x(i) is defined to be sx(1+(i-1)*incx).
 *                         incx must be greater than zero.
 *
 *   precision/hardware  - single/all
 *
 *   reqd. imsl routines - none required
 *
 *   notation            - information on special notation and
 *                         conventions is available in the manual
 *                         introduction or through imsl routine uhelp
 *
 *   copyright           - 1978 by imsl, inc. all rights reserved
 *
 *   warranty            - imsl warrants only that imsl testing has been
 *                         applied to this code. no other warranty,
 *                         expressed or implied, is applicable.
 *
 * ----------------------------------------------------------------------------
 */
int sscal_(f_int *n, double *sa, double *sx, f_int *incx) {
	int		retval = 0;

	/*
	 * First, see if we have anything to do...
	 */
    if (*n > 0) {
    	f_int		i = 0;
		f_int		di = *incx;
		f_int		ns = (*n) * di;
		double	 	a = *sa;

		for (; i < ns; i += di) {
			sx[i] *= a;
   	 	}
	}
    return retval;
}

File: sscal.f
      subroutine sscal (n,sa,sx,incx)
c
c----------------------------------------------------------------------------
c
c   computer            - cdc/single
c
c   latest revision     - january 1, 1978
c
c   purpose             - compute a single precision constant
c                         times a single precision vector
c
c   usage               - call sscal (n,sa,sx,incx)
c
c   arguments    n      - length of vector x. (input)
c                sa     - real scalar. (input)
c                sx     - real vector of length max(n*iabs(incx),1).
c                         (input)
c                         sscal replaces x(i) with sa*x(i) for
c                         i=1,...,n.
c                         x(i) refers to a specific element of sx.
c                         see incx argument description.
c                incx   - displacement between elements of sx. (input)
c                         x(i) is defined to be sx(1+(i-1)*incx).
c                         incx must be greater than zero.
c
c   precision/hardware  - single/all
c
c   reqd. imsl routines - none required
c
c   notation            - information on special notation and
c                         conventions is available in the manual
c                         introduction or through imsl routine uhelp
c
c   copyright           - 1978 by imsl, inc. all rights reserved
c
c   warranty            - imsl warrants only that imsl testing has been
c                         applied to this code. no other warranty,
c                         expressed or implied, is applicable.
c
c----------------------------------------------------------------------------
c
c               specifications for arguments
      integer     incx, n
      double precision   sa
      double precision   sx(1)
c               specifications for local variables
      integer     i, m, mp1, ns
c               first executable statement
      if (n .le. 0) return
      if (incx .eq. 1) go to 10
c               code for increments not eaual to 1
      ns = n*incx
      do 5 i=1,ns,incx
        sx(i) = sa*sx(i)
    5 continue
      return
c               code for increments equal to 1.
c               clean-up loop so remaining vector
c               length is a multiple of 5.
   10 m = n - (n/5)*5
      if (m .eq. 0) go to 20
      do 15 i=1,m
        sx(i) = sa*sx(i)
   15 continue
      if (n .lt. 5) return
   20 mp1 = m + 1
      do 25 i=mp1,n,5
        sx(i) = sa*sx(i)
        sx(i+1) = sa*sx(i+1)
        sx(i+2) = sa*sx(i+2)
        sx(i+3) = sa*sx(i+3)
        sx(i+4) = sa*sx(i+4)
   25 continue
      return
      end
